test/parquet_test.cc

```diff
@@ -17,38 +17,121 @@
+  ArrowArray arr;
+  ICEBERG_ARROW_RETURN_NOT_OK(::arrow::ExportArray(*data, &arr));
+  ICEBERG_RETURN_UNEXPECTED(writer.Write(arr));
+  return writer.Close();
+}
+
+Status WriteTable(std::shared_ptr<::arrow::Array> data,
+                  const WriterOptions& writer_options) {
+  ICEBERG_ASSIGN_OR_RAISE(
+      auto writer, WriterFactoryRegistry::Open(FileFormatType::kParquet, writer_options));
+  return WriteTableInner(*writer, data);
+}
+
+Status ReadTable(std::shared_ptr<::arrow::Array>& out,
+                 const ReaderOptions& reader_options) {
+  ICEBERG_ASSIGN_OR_RAISE(
+      auto reader, ReaderFactoryRegistry::Open(FileFormatType::kParquet, reader_options));
+  ICEBERG_ASSIGN_OR_RAISE(auto read_data, reader->Next());
+
+  if (!read_data.has_value()) {
+    out = nullptr;
+    return {};
+  }
+  auto arrow_c_array = read_data.value();
+
+  ArrowSchema arrow_schema;
+  ICEBERG_RETURN_UNEXPECTED(ToArrowSchema(*reader_options.projection, &arrow_schema));
+  ICEBERG_ARROW_ASSIGN_OR_RETURN(out,
+                                 ::arrow::ImportArray(&arrow_c_array, &arrow_schema));
+  return {};
+}
+
+void DoRoundtrip(std::shared_ptr<::arrow::Array> data, std::shared_ptr<Schema> schema,
+                 std::shared_ptr<::arrow::Array>& out) {
+  std::shared_ptr<FileIO> file_io = arrow::ArrowFileSystemFileIO::MakeMockFileIO();
+  const std::string basePath = "base.parquet";
+
+  auto writer_data = WriterFactoryRegistry::Open(
+      FileFormatType::kParquet, {.path = basePath, .schema = schema, .io = file_io});
+  ASSERT_THAT(writer_data, IsOk())
+      << "Failed to create writer: " << writer_data.error().message;
+  auto writer = std::move(writer_data.value());
+  ASSERT_THAT(WriteTableInner(*writer, data), IsOk());
+
+  ASSERT_THAT(ReadTable(out, {.path = basePath,
+                              .length = writer->length(),
+                              .io = file_io,
+                              .projection = schema}),
+              IsOk());
+
+  ASSERT_TRUE(out != nullptr) << "Reader.Next() returned no data";
+}
+
+}  // namespace
+
+class ParquetReaderTest : public ::testing::Test {
  protected:
   static void SetUpTestSuite() { parquet::RegisterAll(); }
 
   void SetUp() override {
-    TempFileTestBase::SetUp();
-    file_io_ = arrow::ArrowFileSystemFileIO::MakeLocalFileIO();
-    temp_parquet_file_ = CreateNewTempFilePathWithSuffix(".parquet");
+    file_io_ = arrow::ArrowFileSystemFileIO::MakeMockFileIO();
+    temp_parquet_file_ = "parquet_reader_test.parquet";
   }
 
   void CreateSimpleParquetFile() {
+    auto schema = std::make_shared<Schema>(
+        std::vector<SchemaField>{SchemaField::MakeRequired(1, "id", int32()),
+                                 SchemaField::MakeOptional(2, "name", string())});
+
+    ArrowSchema arrow_c_schema;
+    ASSERT_THAT(ToArrowSchema(*schema, &arrow_c_schema), IsOk());
+    auto arrow_schema = ::arrow::ImportType(&arrow_c_schema).ValueOrDie();
```

Comment:
nit: use arrow::ImportSchema

Reply:
Same for below

Reply:
`::arrow::ImportSchema` converts a `ArrowSchema` to `std::shared_ptr<Schema>` and it doesn't seems to be helpful here.

Reply:
My bad. I have mixed Schema with StructType.